### ISSUE TITLE
Lint against single-use lifetime names

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#44752]
Once support for `'_` lands in https://github.com/rust-lang/rust/pull/44691, the next step for https://github.com/rust-lang/rust/issues/44524 is to implement a lint that warns against "single use" lifetime names. 

The idea is that an explicit name like `'a` should only be used (at least in a function or impl) to *link together* two things. Otherwise, you should just use `'_` to indicate that the lifetime is not linked to anything.

Until https://github.com/rust-lang/rust/issues/15872 is closed, we should only lint for single-use lifetime names that are bound in **functions**. Once #15872 is closed, we can also lint against those found in impl headers.

We can detect cases where a lint is valid by modifying [the `resolve_lifetimes` code](https://github.com/rust-lang/rust/blob/3cf28f3002470f61a471c69b869e4f55ae1766f7/src/librustc/middle/resolve_lifetime.rs):

- This code basically walks over the HIR and resolves all lifetime names. 
- It maintains a stack of [scopes](https://github.com/rust-lang/rust/blob/3cf28f3002470f61a471c69b869e4f55ae1766f7/src/librustc/middle/resolve_lifetime.rs#L194-L230) indicating what names are valid.
- [The function `with()`](https://github.com/rust-lang/rust/blob/3cf28f3002470f61a471c69b869e4f55ae1766f7/src/librustc/middle/resolve_lifetime.rs#L781-L803) is used to push new scopes on the stack. It is also given a closure which will execute with the new name bindings in scope.
    - `with()` gets called for impls and other kinds of items [from here](https://github.com/rust-lang/rust/blob/3cf28f3002470f61a471c69b869e4f55ae1766f7/src/librustc/middle/resolve_lifetime.rs#L348-L351); for methods and functions in particular it is called from [`visit_early_late`](https://github.com/rust-lang/rust/blob/3cf28f3002470f61a471c69b869e4f55ae1766f7/src/librustc/middle/resolve_lifetime.rs#L823)).
- Once a name is in scope, [`resolve_lifetime_ref()`](https://github.com/rust-lang/rust/blob/3cf28f3002470f61a471c69b869e4f55ae1766f7/src/librustc/middle/resolve_lifetime.rs#L866) is called to resolve an actual reference to a named lifetime.
    - This could be used, for example, to update some information in the `Scope`, e.g. counting how many times a particular lifetime was referenced.
- Then, before `with()` returns, we could scan the lifetimes and check for those that were only referenced 1 time (or 0 times...) and issue a lint warning.

(There are some directions for how to add a lint under the header "Issuing future compatibility warnings" in [the rustc-bug-fix-procedure page on forge](https://forge.rust-lang.org/rustc-bug-fix-procedure.html) -- we can skip the "future compatibility" parts here.)

**Comment by cramertj** [comment]
>  we should only lint for single-use lifetime names that are bound in functions

Also, I believe, only for lifetime names that appear in argument position. Currently we require explicitly binding output-only lifetimes with a name:
```rust
fn bar<'a>() -> &'a u8 { &5 } // OK
fn bar() -> &'_ u8 { &5 } // ERROR: missing lifetime specifier
```

**Comment by nikomatsakis** [comment]
@cramertj Hmm. I would consider the proper way to declare `bar()`  to probably be `fn bar() -> &'static u8`, really, though there are subtleties involved in rare cases (e.g., around invariance).

**Comment by cramertj** [comment]
@nikomatsakis Yes, `'static` is the right thing to use. I was just pointing out that we shouldn't recommend `'_` here, as it won't work.

**Comment by gaurikholkar** [comment]
Is this up for grabs?

**Comment by cramertj** [comment]
@gaurikholkar Go for it!

**Comment by gaurikholkar** [comment]
Will start working on it

**Comment by nikomatsakis** [comment]
@gaurikholkar hey, just checking in! How's it going? Any blockers?

**Comment by nikomatsakis** [comment]
Some examples:

```rust
fn foo<'x>(_: &'x u32) { }
```

This should lint against `'x` and suggest `&u32` instead.

```rust
struct Foo<'a> { x: &'a u32 }
fn foo<'x>(_: Foo<'x>) { }
```

This should lint against `'x` and suggest `Foo<'_>` instead.

```rust
struct Foo<'a, 'b> { f: &'a &'b u32 }
fn foo<'x, 'y>(foo: Foo<'x, 'y>) -> &'x u32 { foo.f }
```

This should lint against `'y` and suggest `Foo<'x, '_>` instead.

```rust
struct Foo<'a, 'b> { f: &'a &'b u32 }
fn foo<'x, 'y>(foo: Foo<'x, 'y>) -> &'y u32 { foo.f }
```

This should lint against `'x` and suggest `Foo<'_, 'y>` instead.

```rust
fn foo<'x>() -> &'x u32 { &22 }
```

This should not lint, because `'x` appears only in the return type.

```rust
trait Trait<'a> { }
impl<'a, T> Trait<'a> for T { }

fn foo<'x, T>(t: T)
where T: Trait<'x>
{ 
}

fn main() { 
  foo(22);
}
```

This should not lint, because at present `'_` does not work in that position (which we should fix).

**Comment by nikomatsakis** [comment]
OK, let's start with this specific test:

```rust
fn deref<'x>(v: &'x u32) -> u32 {
    *v
}

fn main() { }
```

When we are done, we want to issue a warning, probably like this:

```rust
fn deref<'x>(v: &'x u32) -> u32 {
//       ^^ lifetime name `'x` only used once
    *v
}

fn main() { }
```

In this case, the one use of `'x` is as the operand to a `&`-type, so I think we don't have to say much more than that. If the one use of `'x` were in a struct, we might want to have a hint indicating that it can be replaced with `'_`, but let's leave that stuff to future work.

The first thing we want to do then is to figure out how many times each name is used and where. Now, accounting around lifetimes (e.g., early-bound, late-bound, etc) can get kind of complicated, but luckily we can ignore most of that crap for our purposes. I imagine we would want to add to [the  `LifetimeContext` struct](https://github.com/rust-lang/rust/blob/3cf28f3002470f61a471c69b869e4f55ae1766f7/src/librustc/middle/resolve_lifetime.rs#L164) a new field:

```rust
lifetime_uses: DefIdMap<LifetimeUseSet<'tcx>
```

where a `LifetimeUseSet<'tcx>` is defined like:

```rust
enum LifetimeUseSet<'tcx> {
    One(&'tcx hir::Lifetime),
    Many,
}
```

We don't really need to track more detail than that -- once there are many uses of a lifetime, we don't want to warn about it anymore, so we don't need to track them. Now, when we are resolving a lifetime in [`resolve_lifetime_ref`](https://github.com/rust-lang/rust/blob/3cf28f3002470f61a471c69b869e4f55ae1766f7/src/librustc/middle/resolve_lifetime.rs#L866), we want to update these lifetime use sets. Actually, I think the place to add code is the `insert_lifetime` method, which records a successful lifetime resolution:

https://github.com/rust-lang/rust/blob/3cf28f3002470f61a471c69b869e4f55ae1766f7/src/librustc/middle/resolve_lifetime.rs#L1516-L1518

Here, we want to match on the `def`, which is of [the type `Region`](https://github.com/rust-lang/rust/blob/3cf28f3002470f61a471c69b869e4f55ae1766f7/src/librustc/middle/resolve_lifetime.rs#L40-L46). We want to do something like:

```rust
match def {
    LateBoundAnon(..) | Static => {
        // These are anonymous lifetimes or lifetimes that are not declared.
    }

    Free(_, def_id) | LateBound(_, def_id) | EarlyBound(_, def_id) => {
        // A lifetime declared by the user.
        if !self.lifetime_uses.contains_key(&def_id) {
            self.lifetime_uses.insert(def_id, LifetimeUseSet::One(lifetime_ref));
        } else {
            self.lifetime_uses.insert(def_id, LifetimeUseSet::Many);
        }
    }
}
```

Now we have a record of where each lifetime def was used. The next step will be going over the set of lifetime names defined and warning if they are `LifetimeUseSet::One`. I'm going to stop here, but we can talk about the best way to do that later.

**Comment by gaurikholkar** [comment]
I've been in touch with @nikomatsakis over this. The current status is: I've followed the instructions he has given above and written code for the same locally.

**Comment by nikomatsakis** [comment]
So, if we follow the steps above, we wind up with a map from the def-id of each `LifetimeDef` to either zero (no key), one, or many uses of it. I think maybe the easiest thing way to handle things is to do a second walk of the krate, just looking for lifetime defs, and issuing the warnings. Originally I had thought we could do them as we walk the first time, but now that seems like it would just clutter the code unnecessary.

This is the main function for `resolve_lifetime`:

https://github.com/rust-lang/rust/blob/master/src/librustc/middle/resolve_lifetime.rs#L259-L285

I am imagining that we can basically write a second visitor, `struct LifetimeSingleUseWarnVisitor` or something. Like `LintContext`, it will implement the visitor trait and return `All` for its `nested_visitor_map`:

https://github.com/rust-lang/rust/blob/master/src/librustc/middle/resolve_lifetime.rs#L287-L290

It would only implement *one* method, though, `visit_lifetime_def`:

```
fn visit_lifetime_def(&mut self, lifetime_def: &hir::LifetimeDef) {
    let def_id = self.tcx.hir.as_local_def_id(lifetime_def.id);
    ...
}
```

In there, it will check if the map contains `def_id`. If not, it can issue a warning "lifetime never used". If so, and it contains `One`, it will issue a warning "lifetime used only once". Otherwise, no warning. (Issuing a warning is done by calling `add_lint`, you can grep around for example.)

OK, this is a not *quite* right -- we're going to eventually want to be a bit more selective, since it will depend whether that use is one that could be elided or replaced with `'_`. But let's start there and then narrow it down.

**Comment by nikomatsakis** [comment]
Actually I take that back. Don't use a visitor. Just iterate over the map we built and look things up in the HIR map. For example, given the `DefId` of some lifetime def, we can look up its span in the HIR map by doing

```
let node_id = hir.as_local_node_id(def_id).unwrap(); // guarnateed to be local
let hir_lifetime: &hir::Lifetime = match hir.get(node_id) {
    hir_map::NodeLifetime(l) => l,
    _ => bug!()
};
```

**Comment by nikomatsakis** [comment]
Well, that won't allow you to catch lifetimes used zero times. To do that, we either need a separate visitor, or visit the map during `with` and store dummy nodes, or to visit the entries at the end of `with` and issue warnings.

**Comment by gaurikholkar** [comment]
Track the progress [here](https://github.com/gaurikholkar/rust/commits/single_lifetimes)

**Comment by gaurikholkar** [comment]
```
  |
3 | fn deref<'x>(v: &'x u32) -> u32 {
  |          ^^
  |
note: lint level defined here
 --> /home/nmatsakis/tmp/foo.rs:1:9
  |
1 | #![warn(single_use_lifetime)]
  |         ^^^^^^^^^^^^^^^^^^^
```  
Is what gets generated for now

**Comment by gaurikholkar** [comment]
@nikomatsakis it's compile-fail tests that are failing now

**Comment by nikomatsakis** [comment]
I started trying to record all places where this lint might be needed and document their expected behavior. This is not done, but I have to go, so I'm saving my status here for now. My goal is to make a checklist and try to mentor the remaining improvements.

```rust
#![warn(single_use_lifetime)]
#![allow(dead_code)]
#![allow(unused_variables)]
 
fn a<'a>(x: &'a u32) { // Warning expected
}
 
fn b<'a>() -> &'a u32 { // OK
    &22
}
 
fn c<'a>(x: &'a u32) -> &'a u32 { // OK
    &22
}
 
struct Foo<'f> {
    data: &'f u32
}
 
impl<'f> Foo<'f> { // OK for *now* FIXME(#15872)
    fn d(&self) {
    }

    fn e<'a>(x: &'a u32) { // Warning expected
    }
 
    fn f<'a>() -> &'a u32 { // OK
        &22
    }
 
    fn g<'a>(x: &'a u32) -> &'a u32 { // OK
        &22
    }
}

impl<'f> Foo<'f> { // OK
    fn new(data: &'f u32) -> Self {
        Foo { data }
    }
}

impl<'f> Iterator for Foo<'f> { // OK
    type Item = &'f u32;

    fn next(&mut self) -> Option<Self::Item> {
        None
    }
} 

impl<'f> Foo<'f> { // OK
    fn get(&self) -> &'f u32 {
        self.data
    }
}

fn main() { }
```

**Comment by nikomatsakis** [comment]
Here is an updated check-list of the overall goals:

- [ ] Issue warnings at the right times; [this gist](https://gist.github.com/nikomatsakis/acdeb20b0192e2a1f404d514b3002c59) provides a comprehensive test case.
- [ ] For each warning, issue a suggested fix:
  - For a single-use lifetime `'a` appearing in `&'a T`, suggest `&T`
  - For a single-use lifetime `'a` appearing in any other place, suggest `'_`
  - One challenge: the binder must be removed too
    - cc @estebank -- is it possible to give suggested fixes that make changes to multiple spots at once? I guess that might just be multiple suggestions?

**Comment by gaurikholkar** [comment]
> Well, that won't allow you to catch lifetimes used zero times
This isn't implemented either

**Comment by nikomatsakis** [comment]
Ah, good point. I was wondering if we might want to actually handle the "multiple edits" this way -- that is, we could first suggest removing the `'a` from `&'a u32`, then suggest removing an unused lifetime.

**Comment by nikomatsakis** [comment]
Kind of annoying to have to run `rustfix` to a fixed point though =)

**Comment by nikomatsakis** [comment]
@gaurikholkar I updated the list of warnings

**Comment by estebank** [comment]
> is it possible to give suggested fixes that make changes to multiple spots at once? I guess that might just be multiple suggestions?

The [suggestions machinery is capable of having multiple suggestions at one time, and each suggestions being different `Span`s](https://github.com/rust-lang/rust/blob/master/src/librustc_errors/lib.rs#L63-L107), but there's no method access these features, which means it hasn't been thoroughly tested.

**Comment by nikomatsakis** [comment]
### Issuing warnings at the right times

This is the first step for the mentoring instructions. The lint as is is partially implemented, but it often fires at the wrong times. 

**Step one: Add the tests.** I've created a fairly comprehensive set of test cases in [this gist](https://gist.github.com/nikomatsakis/f13da812c97b8094a9566ca9b3d9677d). The first step then would be to add these tests into your branch, by doing something like the following. This will put all of the tests into `src/test/ui/in-band-lifetimes/single-use-lint`.

```bash
> cd src/test/ui/in-band-lifetimes
> mkdir single-use-lint
> git clone git@gist.github.com:f13da812c97b8094a9566ca9b3d9677d.git tmp
> mv tmp/*rs single-use-lint
> rm -rf tmp
```

(Unfortunately, I forgot when making those tests that we already had a fair number of tests with names like `src/test/ui/in-band-lifetimes/single_use_lifetimes-*rs`. You might just want to remove those tests, or else compare them to the new ones.)

We can then run all the tests by running this command from the main directory:

```bash
> ./x.py test --stage 1 -i src/test/ui --test-args single-use-lint
```

That should run just the new tests and nothing else. You should see various failures. (You can [learn more about rustc's test suite here, in the rustc-guide](https://rust-lang-nursery.github.io/rustc-guide/tests/intro.html); these test are [ui tests](https://rust-lang-nursery.github.io/rustc-guide/tests/adding.html#guide-to-the-ui-tests).) 

**Step two: Examine existing code.** The way that the current lint works is in the `resolve_lifetime` pass.  Basically, for each lifetime that is declared, we track a `LifetimeUseSet` indicating how many times it is used:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L61-L66

These are kept in a map, indexed by the `DefId` from the lifetime definition:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L258

This map is populated in the `insert_lifetime` routine, which is what we invoke when we have resolved some reference to a lifetime. So e.g. if you have a type `&'a u32`, this method has the job of recording which lifetime that `'a` refers to:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L2201

Towards the bottom you will see this code, which upgrades the number of times that the reference is used:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L2228-L2233

**Examine the test failures.** Looking at the test failures, we can group them into two categories:

- Lifetimes that appear in some places should not be warned about, even if they are only used once. For example, [using a lifetime name in a struct](https://gist.github.com/nikomatsakis/f13da812c97b8094a9566ca9b3d9677d#file-one-use-in-struct-rs), or [in a return type](https://gist.github.com/nikomatsakis/f13da812c97b8094a9566ca9b3d9677d#file-one-use-in-fn-return-rs).
- Lifetimes that are never used are not warned.

**Making lifetimes used in particular places issue a warning.** I think what we want to do here is to extend the `LifetimeContext` struct with a new field:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L228

Let's call the flag `track_individual_lifetime_uses: bool`. The idea is that, when this flag is false, `insert_lifetime` is not going to track whether a lifetime is used once or many times: if it finds *any* use, it will treat it as though it is used many times. So we might modify [the `insert_lifetime` code](https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L2228-L2233) as follows:

```rust
if self.track_individual_lifetime_uses && !self.lifetime_uses.contains_key(&def_id) {
    self.lifetime_uses.insert(def_id, LifetimeUseSet::One(lifetime_ref));
} else {
    self.lifetime_uses.insert(def_id, LifetimeUseSet::Many);
}
```

The basic idea is that this field should be `true` whenever a single-use lifetime could be replaced with `'_`. This is actually not true most of the time, so we can initialize the field to `false`. We can then modify the `with` function, which is the function that adds a new lifetime scope, to update the field:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L1247-L1249

`with` should take a new argument (`track_lifetime_uses: bool`); it should save the old value of the flag, modify it, and then restore it before returning.

Usually, we will want to pass `false` as the value of this flag, but there are some exceptions. For example, when checking fn arguments:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L1627-L1637

and function bodies:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L451-L466

Anyway, that should be the rough idea. We'll have to see if it starts to issue warnings at strange times!

**Making lifetimes that are never used issue a warning.** Warnings are actually issued in this loop:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L1273-L1299

The problem here is that, if a lifetime is never used, it will never be added to the map, so we never see it to issue a warning! To fix that, we have to keep a separate set of all lifetimes and iterate over *that* set. Otherwise, we could walk the IR one more time with a separate visitor to find each lifetime definition.

**Comment by tamird** [comment]
@nikomatsakis thanks for the detailed instructions!

The approach you outline doesn't totally work out, I think. Consider [two-uses-in-fn-argument-and-return](https://gist.github.com/nikomatsakis/f13da812c97b8094a9566ca9b3d9677d#file-two-uses-in-fn-argument-and-return-rs) - we want to permit the lifetime here because it has two uses, but we want to pass `false` when calling `self.with` on the output (because lifetimes on the output alone are permitted).

I think we need to always populate `self. lifetime_uses` but only check it when `track_lifetime_uses` is `true`. What do you think?

Other than that, to do the last part (finding lifetimes with zero uses): where would you populate the set of lifetimes?

**Comment by tamird** [comment]
Also, a few of the tests are incoherent:
- [one-use-in-inherent-method-argument](https://gist.github.com/nikomatsakis/f13da812c97b8094a9566ca9b3d9677d#file-one-use-in-inherent-method-argument-rs): the comment says "DO NOT warn" and the test is missing the "run-pass" annotation, but various expected errors are included.
- [one-use-in-inherent-method-return](https://gist.github.com/nikomatsakis/f13da812c97b8094a9566ca9b3d9677d#file-one-use-in-inherent-method-return-rs): same.
- [one-use-in-trait-method-argument](https://gist.github.com/nikomatsakis/f13da812c97b8094a9566ca9b3d9677d#file-one-use-in-trait-method-argument-rs): same.
- [two-uses-in-inherent-method-argument-and-return](https://gist.github.com/nikomatsakis/f13da812c97b8094a9566ca9b3d9677d#file-two-uses-in-inherent-method-argument-and-return-rs): same (does have run-pass, though).

**Comment by nikomatsakis** [comment]
@tamird

> The approach you outline doesn't totally work out, I think

Regarding this... I think actually that [the two-uses-in-fn-argument-and-return test](https://gist.github.com/nikomatsakis/f13da812c97b8094a9566ca9b3d9677d#file-two-uses-in-fn-argument-and-return-rs):

```rust
fn c<'a>(x: &'a u32) -> &'a u32 { // OK: used twice
    &22
}
```

is ok for two independent reasons: first, because the lifetime is used twice, but *also*  because it is used in the return type. So it would be ok to invoke `with` with `false` (while visiting the return type). Hence tests like [one-use-in-inherent-method-return](https://gist.github.com/nikomatsakis/f13da812c97b8094a9566ca9b3d9677d#file-one-use-in-inherent-method-return-rs) still do not warn.

Put another way, I do not yet see the problem. =)

> Other than that, to do the last part (finding lifetimes with zero uses): where would you populate the set of lifetimes?

Hmm. I was thinking about this later. Keeping a set of "all lifetimes" doesn't feel right to me. In fact, the point where we iterate over the lifetimes in order to issue warnings is precisely the point where a set of declared lifetimes are going out of scope:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L1273-L1299

That is, the purpose of the `with` function is to bring new lifetimes into scope. The lifetimes are found on the `scope` parameter, which is of type `Scope<'a>`:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L261-L315

we could iterate over all the lifetimes that were declared in `Scope`. We basically only care about the case where `Scope` is a `Binder`, in which case we can iterate over the `lifetimes` map:

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L268

For each `Region` value found within, we can extract a `DefId` (look for the `/* lifetime_decl */` comments):

https://github.com/rust-lang/rust/blob/c19264fa835a1eca86de4fd2e86a87b3919e57cf/src/librustc/middle/resolve_lifetime.rs#L68-L83

then we can lookup that `DefId` the `lifetime_uses` map. If it is not present, that should be zero uses.

**Comment by nikomatsakis** [comment]
@tamird -- do you think you are going to try and tackle this issue? That would be great! (If not, though, let me know so I can find someone else; I'd like to see this get done.)

**Comment by tamird** [comment]
@nikomatsakis yep, I'm looking at this. Is there any reason to keep `lifetime_uses`? it seems cleaner to combine the usage information into the binder's `lifetimes` map.

**Comment by nikomatsakis** [comment]
@tamird 

> Is there any reason to keep `lifetime_uses`?

If you think you see a cleaner way, go for it!

**Comment by nikomatsakis** [comment]
@tamird btw, just checking -- how goes?  I was away last week for PTO

**Comment by nikomatsakis** [comment]
ping @tamird -- not trying to bug ya', just want to know if you're still working on this. I'm doing my weekly sweep and trying to figure out what I should do to ensure that this moves forward. It's .. mildly high priority, mostly because we'd like to encourage people to try out the in-band lifetimes feature, and for them to do that effectively, this lint ought to be a big help...

**Comment by nikomatsakis** [comment]
First PR: https://github.com/rust-lang/rust/pull/50440

**Comment by joshtriplett** [comment]
Given that the 2018 edition preview included in-band lifetimes, it seems like we need to have this lint available for either the next preview or the 2018 edition stable release.

**Comment by scottmcm** [comment]
Even without in-band lifetimes, I think this is an important idiom lint, so I've added it to an edition milestone to get eyes on it.  Edition folks: please move or remove if you think otherwise.

**Comment by nhynes** [comment]
How do I get this to work with something like serde's `deserialize_with`? In [this specific example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=33bf44b1c931ef5e2c190ac37d649060), the lint complains about the `'de` in `Deserializer<'de>` being single-use, but then replacing it with `'_` nets an `expected named lifetime parameter`.

**Comment by scottmcm** [comment]
Wow, it's been three years...

I still think this is important.  I started looking at how `in_band_lifetimes` in `rustc_mir_transform`, and I immediately found a bunch of examples of unnecessary `'a`s.  Here's just in `coverage/spans.rs`, since it was at the bottom of the error list:

```diff
-    pub fn format(&self, tcx: TyCtxt<'tcx>, mir_body: &'a mir::Body<'tcx>) -> String {
+    pub fn format<'tcx>(&self, tcx: TyCtxt<'tcx>, mir_body: &mir::Body<'tcx>) -> String {

-    pub fn format(&self, tcx: TyCtxt<'tcx>, mir_body: &'a mir::Body<'tcx>) -> String {
+    pub fn format<'tcx>(&self, tcx: TyCtxt<'tcx>, mir_body: &mir::Body<'tcx>) -> String {

-    pub fn format_coverage_statements(
+    pub fn format_coverage_statements<'tcx>(
         &self,
         tcx: TyCtxt<'tcx>,
-        mir_body: &'a mir::Body<'tcx>,
+        mir_body: &mir::Body<'tcx>,
     ) -> String {

-pub(super) fn filtered_statement_span(statement: &'a Statement<'tcx>) -> Option<Span> {
+pub(super) fn filtered_statement_span<'tcx>(statement: &Statement<'tcx>) -> Option<Span> {

-pub(super) fn filtered_terminator_span(terminator: &'a Terminator<'tcx>) -> Option<Span> {
+pub(super) fn filtered_terminator_span<'tcx>(terminator: &Terminator<'tcx>) -> Option<Span> {
```

**Comment by jackh726** [comment]
Okay, I've created the `F-lint-single_use_lifetiems` label and tagged the 6 relevant issues open for this.

The examples in the "current status" section at the top all warn and have a suggestion, except for the unused lifetime, which falls under the `unused_lifetime` lint.

Marking S-tracking-impl-complete, given that there are several linked issues that need to be fixed.

**Comment by est31** [comment]
With #96833 merged, there has been progress on this issue recently. The number of false positives is now way down, but there's still one remaining, according to the PR's description.

**Comment by cjgillot** [comment]
@est31 what is the remaining false positive?

**Comment by est31** [comment]
@cjgillot I was referring to:

> Remaining false positive: single-use lifetimes in argument-position impl-trait.
I'm waiting for #96529 to be fixed to have a clean and proper solution here.

Or is the false positive fixed now?

**Comment by cjgillot** [comment]
Indeed, the remaining false positive is:
```rust
trait Foo<'a> {}

#[warn(single_use_lifetimes)]
fn foo<'a>(_: impl Foo<'a>) {}
```

Where the lifetime cannot be elided.  There is an inconsistency here between plain functions and async functions (which allow elision). I suggested to allow elision in https://github.com/rust-lang/rust/pull/97720.

If that change is accepted, there won't be a false positive any more, and the suggestion will be correct.

**Comment by bstrie** [comment]
I think the following might be an unaddressed false positive:

```rust
    pub fn get_json<T>(&self, limit: u64) -> Result<(Meta, T)>
    where
        for<'de> T: Deserialize<'de>,
```
```
error: lifetime parameter `'de` only used once
   --> crates/client/src/entity.rs:182:13
    |
182 |         for<'de> T: Deserialize<'de>,
    |             ^^^                 --- ...is used only here
    |             |
    |             this lifetime...
```
I'm not aware of any way to use anonymous lifetimes to avoid the single-use lifetime here.

**Comment by bstrie** [comment]
Another false positive:
```rust
fn deserialize<'de, D: Deserializer<'de>>(deserializer: D) -> Result<Mime, D::Error> {
```
```
error: lifetime parameter `'de` only used once
  --> crates/type/src/meta.rs:31:16
   |
31 | fn deserialize<'de, D: Deserializer<'de>>(deserializer: D) -> Result<Mime, D::Error> {
   |                ^^^ this lifetime... --- ...is used only here
```

Are anonymous lifetimes supposed to work here? The error:

```
error[E0637]: `'_` cannot be used here
  --> crates/type/src/meta.rs:31:32
   |
31 | fn deserialize<D: Deserializer<'_>>(deserializer: D) -> Result<Mime, D::Error> {
   |                                ^^ `'_` is a reserved lifetime name
```

**Comment by cjgillot** [comment]
@bstrie, those cases are correctly handled by beta 1.63.0 https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=4fd54ae5a4d4a3a936054787f6655b7b

**Comment by clarfonthey** [comment]
This appears to ignore the fact that lifetimes inside `impl Trait` are currently unstable. See:

```rust
#![warn(single_use_lifetimes)]
fn test<'a>(val: &(impl 'a + Iterator)) {}
```

This actually suggests removing the lifetime, which in turn is suggested to be added back with:

```rust
#![warn(single_use_lifetimes)]
fn test(val: &(impl '_ + Iterator)) {}
```

**Comment by joshtriplett** [comment]
Summarizing the current state of this:

- https://github.com/rust-lang/rust/pull/97720 was merged, but it requires the unstable feature `anonymous_lifetime_in_impl_trait`
- `anonymous_lifetime_in_impl_trait` doesn't have a tracking issue, but https://github.com/rust-lang/rust/pull/107378 proposes stabilization of it.
- https://github.com/rust-lang/rust/pull/107378 is blocked on a concern about interaction with GATs. There are some suggestions to reduce the scope of what's being stabilized, but I wasn't clear on the state of that.
- `anonymous_lifetime_in_impl_trait` also has an ICE: https://github.com/rust-lang/rust/issues/124340

I *think* to make forward progress on this:
- We need a fix for the ICE, and
- Either:
  - Lang needs to discuss the GAT issue and reach a consensus, or
  - The stabilization scope needs to reduce to not include that, and lang needs to come to a consensus on the reduced-scope stabilization

**Comment by zacknewman** [comment]
Is the following a "false positive"?

Existing code is the following:

```rust
pub struct Foo<'a>(pub &'a str);
impl<'a> Foo<'a> {
    // The below lint causes a compilation failure.
    #[deny(single_use_lifetimes)]
    pub fn new<'b: 'a>(x: &'b str) -> Self {
        Self(x)
    }
}
```

Code that uses it looks like below:

```rust
use either::Either;
    
fn example() {
    Foo::new::<'static>("");
}

fn example2<I>(s: &str, i: I, b: bool) -> impl Iterator<Item = Foo<'_>>
where
    I: Iterator<Item = &'static str>,
{
    if b {
        Either::Left(Some(Foo::new(s)).into_iter())
    } else {
        Either::Right(i.map(Foo::new))
    }
}
```

If `Foo::new` is changed to accommodate the lint, then `example1` and `example2` do not compile.

**Comment by GrigorenkoPV** [comment]
> Is the following a "false positive"?

Looks like a false positive to me, yes. Seems like this was also reported in #153836.
